File: ex9/Atividade.java
import java.util.ArrayList;
import java.util.List;


// ============================================================
// EXCEÇÕES
// POR FAVOR, NÃO MODIFICAR
// ============================================================

class SensorIndisponivelException extends RuntimeException {

    public SensorIndisponivelException(String mensagem) {
        super(mensagem);
    }
}


class MedidorJaRegistradoException extends RuntimeException {

    public MedidorJaRegistradoException(String mensagem) {
        super(mensagem);
    }
}


class MedidorNaoEncontradoException extends RuntimeException {

    public MedidorNaoEncontradoException(String mensagem) {
        super(mensagem);
    }
}


// ============================================================
// CONTRATO UTILIZADO PELA ESTAÇÃO
// POR FAVOR, NÃO MODIFICAR
// ============================================================

interface Medidor {

    String getId();

    double medir();
}


// ============================================================
// COMPONENTES EXTERNOS
// POR FAVOR, NÃO MODIFICAR
// ============================================================

class TermometroAntigo {

    private final String codigo;
    private final double temperaturaFahrenheit;

    public TermometroAntigo(
            String codigo,
            double temperaturaFahrenheit
    ) {
        this.codigo = codigo;
        this.temperaturaFahrenheit = temperaturaFahrenheit;
    }

    public String codigo() {
        return codigo;
    }

    public double lerFahrenheit() {
        return temperaturaFahrenheit;
    }
}


class SensorDigital {

    private final String serial;
    private final double temperaturaCelsius;
    private final boolean operacional;

    public SensorDigital(
            String serial,
            double temperaturaCelsius,
            boolean operacional
    ) {
        this.serial = serial;
        this.temperaturaCelsius = temperaturaCelsius;
        this.operacional = operacional;
    }

    public String serial() {
        return serial;
    }

    public boolean estaOperacional() {
        return operacional;
    }

    public double obterLeitura(String unidade) {

        if (!"C".equals(unidade)) {
            throw new IllegalArgumentException(
                    "Unidade não suportada"
            );
        }

        return temperaturaCelsius;
    }
}


// ============================================================
// A PARTIR DAQUI É SUA IMPLEMENTAÇÃO...
// ============================================================

class TermometroAdapter implements Medidor {

    private final TermometroAntigo termometro;

    public TermometroAdapter(TermometroAntigo termometro) {
        this.termometro = termometro;
    }

    @Override
    public String getId() {
        return null;
    }

    @Override
    public double medir() {
        return 0.0;
    }
}


class SensorDigitalAdapter implements Medidor {

    private final SensorDigital sensor;

    public SensorDigitalAdapter(SensorDigital sensor) {
        this.sensor = sensor;
    }

    @Override
    public String getId() {
        return null;
    }

    @Override
    public double medir() {
        return 0.0;
    }
}


class CentralMonitoramento {

    private CentralMonitoramento() {
    }

    public static CentralMonitoramento getInstance() {
        return null;
    }

    public void registrar(Medidor medidor) {
    }

    public boolean remover(String identificador) {
        return false;
    }

    public int quantidade() {
        return 0;
    }

    public double medir(String identificador) {
        return 0.0;
    }

    public double temperaturaMedia() {
        return 0.0;
    }

    public List<String> sensoresEmAlerta(double limite) {
        return new ArrayList<>();
    }
}


// Classe mantida apenas para permitir a entrega como Atividade.java.
// Não é necessário implementar main.
public class Atividade {
}

File: ex9/atividade.py
from abc import ABC, abstractmethod


class SensorIndisponivelException(Exception):
    pass


class MedidorJaRegistradoException(Exception):
    pass


class MedidorNaoEncontradoException(Exception):
    pass


class Medidor(ABC):

    @abstractmethod
    def get_id(self):
        pass

    @abstractmethod
    def medir(self):
        pass


# ============================================================
# COMPONENTES EXTERNOS
# POR FAVOR, NÃO MODIFICAR
# ============================================================

class TermometroAntigo:

    def __init__(self, codigo, temperatura_fahrenheit):
        self.__codigo = codigo
        self.__temperatura_fahrenheit = temperatura_fahrenheit

    def codigo(self):
        return self.__codigo

    def ler_fahrenheit(self):
        return self.__temperatura_fahrenheit


class SensorDigital:

    def __init__(self, serial, temperatura_celsius, operacional=True):
        self.__serial = serial
        self.__temperatura_celsius = temperatura_celsius
        self.__operacional = operacional

    def serial(self):
        return self.__serial

    def esta_operacional(self):
        return self.__operacional

    def obter_leitura(self, unidade):
        if unidade != "C":
            raise ValueError("Unidade não suportada")

        return self.__temperatura_celsius


# ============================================================
# A PARTIR DAQUI É SUA IMPLEMENTAÇÃO...
# ============================================================

class TermometroAdapter(Medidor):

    def __init__(self, termometro):
        self.termometro = termometro

    def get_id(self):
        return self.termometro.codigo()

    def medir(self):
        return (self.termometro.ler_fahrenheit() - 32) * 5 / 9


class SensorDigitalAdapter(Medidor):

    def __init__(self, sensor):
        self.sensor = sensor

    def get_id(self):
        return self.sensor.serial()

    def medir(self):
        
        if (self.sensor.esta_operacional() == False):
            raise SensorIndisponivelException

        return self.sensor.obter_leitura("C")


class CentralMonitoramento:
    
    _instance = None
    
    def __init__(self):
        self.medidores = {}

    @classmethod
    def get_instance(cls):
        if (cls._instance is None):
            cls._instance = CentralMonitoramento()
        return cls._instance


    def registrar(self, medidor):
        if (medidor.get_id() in self.medidores):
            raise MedidorJaRegistradoException
        self.medidores[medidor.get_id()] = medidor
        

    def remover(self, identificador):
        if (identificador not in self.medidores):
            return False
        self.medidores.pop(identificador)
        return True

    def quantidade(self):
        return len(self.medidores)

    def medir(self, identificador):
        if (identificador not in self.medidores):
            raise MedidorNaoEncontradoException
        return self.medidores[identificador].medir()

    def temperatura_media(self):
        temp = 0.0
        count = 0
        
        for medidor in self.medidores.values():
            try:
                temp += medidor.medir()
                count +=1
            except SensorIndisponivelException:
                continue
        return temp / count if count > 0 else 0.0

    def sensores_em_alerta(self, limite):
    
        identificadores = []
        
        for medidor in self.medidores.values():
                    try:
                        if (medidor.medir() > limite):
                            identificadores.append(medidor.get_id())
                    except SensorIndisponivelException:
                        continue    
        return identificadores